### PR TITLE
Use random backoff for retries in cloud-cidr-allocator

### PR DESCRIPTION
**Pull request by shyamjvs**
Ref https://github.com/kubernetes/kubernetes/pull/68084#issuecomment-417651247
/cc @wojtek-t 

```release-note
NONE
```

### CODE REVIEW

File: pkg/controller/nodeipam/ipam/cidr_allocator.go [hunk 1]
```diff
@@ -70,11 +70,11 @@ const (
 	// cidrUpdateRetries is the no. of times a NodeSpec update will be retried before dropping it.
 	cidrUpdateRetries = 3
 
-	// updateRetryTimeout is the time to wait before requeing a failed node for retry
-	updateRetryTimeout = 250 * time.Millisecond
+	// updateRetryTimeoutMilliseconds is the time to wait before requeing a failed node for retry
+	updateRetryTimeoutMilliseconds = 250
```
Comment (wojtek-t) [inline pkg/controller/nodeipam/ipam/cidr_allocator.go]: Why this change - I think having it as a duration is more intuitive.

File: pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go [hunk 2]
```diff
@@ -206,14 +207,14 @@ func (ca *cloudCIDRAllocator) retryParams(nodeName string) (bool, time.Duration)
...
-	timeout := updateRetryTimeout
-	for i := 0; i < count && timeout < maxUpdateRetryTimeout; i++ {
+	timeout := updateRetryTimeoutMilliseconds
+	for i := 0; i < count && timeout < maxUpdateRetryTimeoutMilliseconds; i++ {
 		timeout *= 2
 	}
-	if timeout > maxUpdateRetryTimeout {
-		return maxUpdateRetryTimeout
+	if timeout > maxUpdateRetryTimeoutMilliseconds {
+		timeout = maxUpdateRetryTimeoutMilliseconds
```
Comment (wojtek-t) [inline pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go]: And then you here would need:
```
time.Duration(rand.Int31n(timeout/time.Millisecond)) * time.Millisecond
```
But I still think it's cleaner.

File: pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go [hunk 3]
```diff
@@ -206,14 +207,14 @@ func (ca *cloudCIDRAllocator) retryParams(nodeName string) (bool, time.Duration)
...
 	}
-	return timeout
+	return time.Duration(rand.Int31n(timeout)) * time.Millisecond
```
Comment (wojtek-t) [inline pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go]: I think that it's not how we should do that. The expected value of that is 1/2 of what we have now. Also the variation here is potentially too big.

Instead of that, I think we should make it more like "uniformly distributed between [1/2*timeout, timeout].
Or timeout + some random number?
Or uniformly distributed between [0.75*timeout, 1.25*timeout] (I think it's what Jitter does..)

But maybe good to get @bowei `s opinion too.

File: pkg/controller/nodeipam/ipam/cidr_allocator.go (hunk 1 above)
Comment (shyamjvs) [reply pkg/controller/nodeipam/ipam/cidr_allocator.go]: Done.

File: pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go (hunk 2 above)
Comment (shyamjvs) [reply pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go]: I agree. Changed it.

File: pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go (hunk 3 above)
Comment (shyamjvs) [reply pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go]: I changed it to `[0.5,1.5]*timeout` to keep the expectation same. The variance I think should be high so that we can actually spread out the calls enough.
WIll wait for Bowei.

**Comment by bowei** [comment]
thx 

/lgtm 
/approve

**Comment by dims** [comment]
/milestone v1.12

This is a follow up to a PR targetting 1.12 to fix a problem.
